Lib/test/test_threading.py

```diff
@@ -2103,6 +2103,61 @@ def test__all__(self):
         support.check__all__(self, threading, ('threading', '_thread'),
                              extra=extra, not_exported=not_exported)
 
+    def test_set_name(self):
+        try:
+            get_name = _thread._get_name
+            _thread.set_name
+        except AttributeError:
+            self.skipTest("need thread._get_name() and thread.set_name()")
+
+        def work():
+            nonlocal work_name
+            work_name = get_name()
+
+        # name not too long to fit into Linux 15 bytes limit
+        name = "CustomName"
+        tests = [(name, name)]
+
+        # Test non-ASCII short name
+        name = "namé€"
+        try:
+            os.fsencode(name)
+        except UnicodeEncodeError:
+            # name cannot be encoded to the filesystem encoding
+            pass
```

Comment:
What should happen in this case?

IMO, if we fail to set the native thread name, creating the thread should still succeed. Something like:

```suggestion
            # name cannot be encoded to the filesystem encoding
            tests.append((name, None))
```
with later code to skip the assert on `None`, but ensure that the thread itself starts.


I ran the tests with this change and with `\N{SPOOL OF THREAD}` instead of `é€`, and with `LC_ALL=en_US.iso885915`, and got a scary error message:

```
Exception ignored in thread started by <bound method Thread._bootstrap of <Thread(xxxxxxxxxxxxxx\U0001f9f5, initial 140042635835072)>>:
Traceback (most recent call last):
  File "/home/encukou/dev/cpython/Lib/threading.py", line 1016, in _bootstrap
    self._bootstrap_inner()
  File "/home/encukou/dev/cpython/Lib/threading.py", line 1051, in _bootstrap_inner
    self._delete()
  File "/home/encukou/dev/cpython/Lib/threading.py", line 1056, in _delete
    del _active[get_ident()]
KeyError: 140042635835072
```

Perhaps trusting  Argument Clinic with `PyUnicode_FSConverter` is not a good idea :(

Reply:
Oops. I modified the PR to not set the name if the name cannot be encoded to the filesystem encoding.

I also enhanced the tests to test more cases.